### PR TITLE
Namematcher improvements

### LINKED ISSUE #1185: Gradle doesn't report abbreviated task name as ambiguous when task name has numbers
Given this build file:
```
task unique1()
task uniqueA()
```

Running with this command-line should produce an ambiguous task selection error between unique1 and uniqueA.
```
gradle unique
```

Instead, we select unique1.   Tested with 1.0, 2.0, 2.2.1 and master (2.4 nightly). 

Migrated from: https://issues.gradle.org/browse/GRADLE-3258

### PR DESCRIPTION
**Pull request by ov7a**
Fixes #1185, includes tests updates and refactorings

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/util/internal/NameMatcher.java [hunk 1]
```diff
@@ -145,56 +148,46 @@ public String find(String pattern, Collection<String> items) {
...
-            } else {
-                // Manually extend character class instead of properties
-                // due to a known issue in regex below Java 15
-                // (https://bugs.openjdk.org/browse/JDK-8239887).
-                // \u00DF-\u00F6(ß-ö) - Covers extended Latin lowercase letters including German
-                //  Eszett and various accented characters.
-                // \u00F8-\u00FF(ø-ÿ) - Covers extended Latin lowercase letters including ø,
-                //  accented vowels, and the thorn.
-                // Note: this approach is not as comprehensive as properties.
-                builder.append("[a-z\\u00DF-\\u00F6\\u00F8-\\u00FF\\p{Digit}]*");
```
Comment (ljacomet) [inline subprojects/core/src/main/java/org/gradle/util/internal/NameMatcher.java:171]: 🤔 How come this is no longer needed?

I ran the test locally with Java 8, 11, 17 and 22. They all pass 🎉 
Do you know which change makes https://bugs.openjdk.org/browse/JDK-8239887 no longer applicable?

Context: #23580

**Review by ljacomet — APPROVED** [review]
Nice, and with simplified code!

File: subprojects/core/src/main/java/org/gradle/util/internal/NameMatcher.java (hunk 1 above)
Comment (ov7a) [reply subprojects/core/src/main/java/org/gradle/util/internal/NameMatcher.java:171]: It should be `| Pattern.UNICODE_CASE`:
https://github.com/gradle/gradle/blob/e54b94dd0b6732ea92bd5b967c0836630c90ae49/subprojects/core/src/main/java/org/gradle/util/internal/NameMatcher.java#L85

**Comment by ov7a** [comment]
@bot-gradle APT

**Comment by ov7a** [comment]
@bot-gradle test APT

**Review by cobexer — APPROVED** [review]
LGTM!

File: .teamcity/performance-test-durations.json [hunk 2]
```diff
@@ -639,7 +639,7 @@
     "linux" : 204
   } ]
 }, {
-  "scenario" : "org.gradle.performance.regression.nativeplatform.NativeBuildDependentsPerformanceTest.run libA0:buildDependentsLibA0",
+  "scenario" : "org.gradle.performance.regression.nativeplatform.NativeBuildDependentsPerformanceTest.run libA0:buildDependentsLibA00",
```
Comment (ljacomet) [inline .teamcity/performance-test-durations.json:642]: As discussed, this is worth adding in the upgrade guide

Comment (ov7a) [reply .teamcity/performance-test-durations.json:642]: https://github.com/gradle/gradle/pull/30538

**Pull request merged**